### ISSUE TITLE
Remove the non JS fallbacks

### ISSUE THREAD

**Issue by leofeyer** [contao/core-bundle#583]
I would like to remove the non JavaScript fallback routines of the widgets (list wizard, table wizard etc.) and the back end views (`&tg=` etc.). Since the web does not really work without JS these days, the routines have become a burden.

@contao/developers Any objections?

**Comment by ausi** [comment]
I’d vote against removing them. IMO the basic functionality should work without JavaScript. Also because there can be many different reasons why the JavaScript version might not work, like failing to load the JS file (bad connection, proxy or ad-blocker), a syntax error or a browser bug.

**Comment by aschempp** [comment]
I would not remove them either, but it very much depends on whether the backend works without JS in general.

**Comment by leofeyer** [comment]
> I would not remove them either, but it very much depends on whether the backend works without JS in general.

**Comment by leofeyer** [comment]
> I would not remove them either, but it very much depends on whether the backend works without JS in general.

It does not really. But I will make a more detailed list of what works and what does not work tomorrow.

**Comment by leofeyer** [comment]
As promised, here is a list of things that change when JS is disabled.

**TL;DR** The back end is not usable without JS, because the pickers do not work (page picker, file picker).

| Function | Issues | Fixable | Show stopper |
| --- | --- | --- | --- |
| Modal dialogs | Modal dialogs open a new page instead of a modal dialog | No | No |
| Page/file selector | The selectors open a new page instead of a modal dialog and since the "apply" button is within the modal dialog, you cannot select anything | No | **Yes** |
| Important part | The important part needs to be defined using 4 text fields for width, height, X and Y | No | No |
| Checkbox wizard | Opening and closing sections does not work | Yes | No |
| Palettes | Palettes are always open and cannot be collapsed | No | No |
| Advanced widgets | Obviously things like drag sorting, date picker, chosen, TinyMCE or ACE do not work | No | No |
| Wizards | The "move up/down" functions of all widgets are broken | Yes | No |
| Subpalettes | Subpalettes are not loaded on click (submit on change) but only after the record has been saved | No | No |

Also, the green eye icon, which toggles the visibility of a record, has a non JS fallback, although there is another fallback routine already (namely the checkbox in the "edit" mask).

**Comment by aschempp** [comment]
I suppose the page/file selectors could be fixed, but I don't know if that's worth the efforts. Yesterday I stumbled upon the following article: https://gds.blog.gov.uk/2013/10/21/how-many-people-are-missing-out-on-javascript-enhancement/

If we really drop support for non-JS, we must add a warning to the backend (e.g. below the header bar) that clearly states the system is not useable without JS.

**Comment by leofeyer** [comment]
> I suppose the page/file selectors could be fixed

How would you fix it?

**Comment by aschempp** [comment]
> How would you fix it?

Using a form and submit button.

**Comment by leofeyer** [comment]
That's ugly but it would work I suppose.

> If we really drop support for non-JS, we must add a warning to the backend (e.g. below the header bar) that clearly states the system is not useable without JS.

This would be my preferred solution. 😄

**Comment by leofeyer** [comment]
Also, please note that we would get rid of a lot of ballast such as the `array_move_up()` and `array_insert()` etc. functions, which are currently defined in the `functions.php` file and which we wanted to get rid of anyway.

**Comment by aschempp** [comment]
But we can't drop them until Contao 5. And in Contao 5 we could as well drop non-JS or move to React/Angular anyway…

**Comment by leofeyer** [comment]
True, but no longer using them is the first step to dropping them, isn't it?

**Comment by ausi** [comment]
> If we really drop support for non-JS, we must add a warning to the backend (e.g. below the header bar) that clearly states the system is not useable without JS.

I don’t like that. I’m glad that the times of “Optimized for IE6 1024x768 with JS turned on” are gone.

**Comment by leofeyer** [comment]
The message would only be shown if JS is disabled of course. There is an "You have to allow cookies to use Contao" message in the install tool already. 😄

**Comment by leofeyer** [comment]
Changed in a2d838709e2a4b21b71b7780cb6d3228f648686a. I have mostly just removed the up and down arrows in favour of the drag handle and added a notice to the log in screen. I have also fixed all remaining non JS fallback routines.

**Comment by leofeyer** [comment]
I have deprecated the custom array helper functions in a7ea42a716eafc061751ed637a1b28be8ab95421.

**Comment by ausi** [comment]
But AFAIR the up and down arrows also worked with JavaScript enabled for blind or keyboard users. How do I move things now if I don’t have a mouse?

**Comment by leofeyer** [comment]
I don't think they did.
